builder/builder-next/builder.go

```diff
@@ -551,17 +554,26 @@ func (j *buildJob) SetUpload(ctx context.Context, rc io.ReadCloser) error {
 }
 
 // toBuildkitExtraHosts converts hosts from docker key:value format to buildkit's csv format
-func toBuildkitExtraHosts(inp []string) (string, error) {
+func toBuildkitExtraHosts(inp []string, hostGatewayIP net.IP) (string, error) {
 	if len(inp) == 0 {
 		return "", nil
 	}
 	hosts := make([]string, 0, len(inp))
 	for _, h := range inp {
 		parts := strings.Split(h, ":")
-
-		if len(parts) != 2 || parts[0] == "" || net.ParseIP(parts[1]) == nil {
+		if len(parts) != 2 || parts[0] == "" || (parts[1] != opts.HostGatewayName && net.ParseIP(parts[1]) == nil) {
 			return "", errors.Errorf("invalid host %s", h)
 		}
+		// If the IP Address is a string called "host-gateway", replace this
+		// value with the IP address stored in the daemon level HostGatewayIP
+		// config variable
+		if parts[1] == opts.HostGatewayName {
+			gateway := hostGatewayIP.String()
+			if gateway == "" {
+				return "", fmt.Errorf("unable to derive the IP value for host-gateway")
+			}
+			parts[1] = gateway
+		}
```

Comment:
Oh! Looks like I missed this one in https://github.com/moby/moby/pull/44381 (perhaps we should change it to `strings.Cut()` here as well;

```suggestion
		host, ip, ok := strings.Cut(h, ":")
		if !ok || host == "" || ip == "" {
			return "", errors.Errorf("invalid host %s", h)
		}
		// If the IP Address is a "host-gateway", replace this value with the
		// IP address stored in the daemon level HostGatewayIP config variable.
		if ip == opts.HostGatewayName {
			gateway := hostGatewayIP.String()
			if gateway == "" {
				return "", fmt.Errorf("unable to derive the IP value for host-gateway")
			}
			ip = gateway
		} else if net.ParseIP(ip) == nil {
			return "", fmt.Errorf("invalid host %s: invalid IP value: %s", host, ip)
		}
```

Reply:
@crazy-max could you have a look at my suggestions if they look good to you?